### DISCUSSION TITLE
Newest release and a question about a little bit of memory

### DISCUSSION THREAD

**Discussion by mcctuxic** [tabemann/zeptoforth#125] [General]
Hi Travis,

thanks a lot for the newest release - it is already flashed on my two boards ! :)

I need to send 16 bytes 'en bloc' via SPI to another board.
While trying to understand pool and heap and temporay buffers I got lost somehow.

The memory of those 16 bytes must be sequential - no gaps inbetween.
Each byte of the memory will be read/written by the program very often and sent
via SPI after each modification.
The memory should be in RAM, the program is compiled to flash.

What is the most (zepto-) forthistic yet simple way to accomplish this?

Cheers!
Tuxic

**Comment by tabemann** [comment]
On Thu, Aug 1, 2024 at 10:13 AM mcctuxic ***@***.***> wrote:

> Hi Travis,
>
> thanks a lot for the newest release - it is already flashed on my two
> boards ! :)
>
> I will probably soon make another release, enabling the user to set the
terminal height manually rather than having MORE-* query the terminal for
it (which will still be the default behavior), in case the user wants to
use the MORE-* words with things such as zeptocom.js which do not support
such querying. I will also be looking into creating a general-purpose MORE
word which can be used with things other than just displaying words. (This
will be harder to implement than MORE-WORDS and like, though.)

> I need to send 16 bytes 'en bloc' via SPI to another board.
> While trying to understand pool and heap and temporay buffers I got lost
> somehow.
>
> The memory of those 16 bytes must be sequential - no gaps inbetween.
> Each byte of the memory will be read/written by the program very often and
> sent
> via SPI after each modification.
> The memory should be in RAM, the program is compiled to flash.
>
> What is the most (zepto-) forthistic yet simple way to accomplish this?
>
> Cheers!
> Tuxic
>
Use MY-BUFFER 16 MY-SPI SPI::BUFFER>SPI where MY-BUFFER is the address of
your buffer in RAM and MY-SPI is the index of the SPI device you are using.
Assuming your SPI device is configured to send bytes of data this will send
16 bytes starting at MY-BUFFER to the SPI device MY-SPI while discarding
the bytes sent the opposite direction via SPI.

Travis

**Comment by mcctuxic** [comment]
On 08/01 10:41, tabemann wrote:
> On Thu, Aug 1, 2024 at 10:13 AM mcctuxic ***@***.***> wrote:
> 
> > Hi Travis,
> >
> > thanks a lot for the newest release - it is already flashed on my two
> > boards ! :)
> >
> > I will probably soon make another release, enabling the user to set the
> terminal height manually rather than having MORE-* query the terminal for
> it (which will still be the default behavior), in case the user wants to
> use the MORE-* words with things such as zeptocom.js which do not support
> such querying. I will also be looking into creating a general-purpose MORE
> word which can be used with things other than just displaying words. (This
> will be harder to implement than MORE-WORDS and like, though.)

...this will also become handy in cases one don't want to fill the whole
screen with words - very neat ! :)

> 
> > I need to send 16 bytes 'en bloc' via SPI to another board.
> > While trying to understand pool and heap and temporay buffers I got lost
> > somehow.
> >
> > The memory of those 16 bytes must be sequential - no gaps inbetween.
> > Each byte of the memory will be read/written by the program very often and
> > sent
> > via SPI after each modification.
> > The memory should be in RAM, the program is compiled to flash.
> >
> > What is the most (zepto-) forthistic yet simple way to accomplish this?
> >
> > Cheers!
> > Tuxic
> >
> Use MY-BUFFER 16 MY-SPI SPI::BUFFER>SPI where MY-BUFFER is the address of
> your buffer in RAM and MY-SPI is the index of the SPI device you are using.
> Assuming your SPI device is configured to send bytes of data this will send
> 16 bytes starting at MY-BUFFER to the SPI device MY-SPI while discarding
> the bytes sent the opposite direction via SPI.
> 
> Travis

Thank you for your support, Travis! :) 
One problem was for me, what method
of acquiring/allocating 16 byte of memory
would be the best? The memory will be reused...

**Comment by tabemann** [comment]
On Thu, Aug 1, 2024 at 8:38 PM mcctuxic ***@***.***> wrote:

> On 08/01 10:41, tabemann wrote:
> > On Thu, Aug 1, 2024 at 10:13 AM mcctuxic ***@***.***> wrote:
> >
> > > Hi Travis,
> > >
> > > thanks a lot for the newest release - it is already flashed on my two
> > > boards ! :)
> > >
> > > I will probably soon make another release, enabling the user to set the
> > terminal height manually rather than having MORE-* query the terminal for
> > it (which will still be the default behavior), in case the user wants to
> > use the MORE-* words with things such as zeptocom.js which do not support
> > such querying. I will also be looking into creating a general-purpose
> MORE
> > word which can be used with things other than just displaying words.
> (This
> > will be harder to implement than MORE-WORDS and like, though.)
>
> ...this will also become handy in cases one don't want to fill the whole
> screen with words - very neat ! :)
>

I have implemented a new general-purpose pager that can be gotten (it is
not in a build yet) at
https://github.com/tabemann/zeptoforth/blob/master/src/common/forth/more.fs.
Note that I have not specified manually selecting a terminal size because I
found it did not do what I wanted it to do, and it would either be
misleading or, if I managed to make it do what was expected,
overcomplicated implementation-wise.


> >
> > > I need to send 16 bytes 'en bloc' via SPI to another board.
> > > While trying to understand pool and heap and temporay buffers I got
> lost
> > > somehow.
> > >
> > > The memory of those 16 bytes must be sequential - no gaps inbetween.
> > > Each byte of the memory will be read/written by the program very often
> and
> > > sent
> > > via SPI after each modification.
> > > The memory should be in RAM, the program is compiled to flash.
> > >
> > > What is the most (zepto-) forthistic yet simple way to accomplish this?
> > >
> > > Cheers!
> > > Tuxic
> > >
> > Use MY-BUFFER 16 MY-SPI SPI::BUFFER>SPI where MY-BUFFER is the address of
> > your buffer in RAM and MY-SPI is the index of the SPI device you are
> using.
> > Assuming your SPI device is configured to send bytes of data this will
> send
> > 16 bytes starting at MY-BUFFER to the SPI device MY-SPI while discarding
> > the bytes sent the opposite direction via SPI.
> >
> > Travis
>
> Thank you for your support, Travis! :)
> One problem was for me, what method
> of acquiring/allocating 16 byte of memory
> would be the best? The memory will be reused...


If you want to reuse the space I would do something like:

: my-word
  16 [: { my-buffer }
    my-buffer fill-buffer-with-data
    my-buffer 16 my-spi spi::buffer>spi
  ;] with-aligned-allot
;

WITH-ALIGNED-ALLOT will temporarily ALLOT the space in a cell-aligned
fashion in the current task's dictionary, and then un-ALLOT it when the xt
passed in returns, even if an exception occurs (and if it does, it
re-raises it afterwards).

Travis
